### PR TITLE
dump: Rewrite Linux ACL handling

### PR DESCRIPTION
**Pull request by greatroar**
What does this PR change? What problem does it solve?
-----------------------------------------------------

Rewrites the ACL handling code in restic dump. The old version was taken from an MPL-licensed library. This is a cleanroom implementation. The code is shorter and it's now explicit that only Linux ACLs are supported.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

No.

Checklist
---------

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added tests for all code changes.
- [ ] I have added documentation for relevant changes (in the manual).
- [ ] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

### CODE REVIEW

File: internal/dump/acl.go [hunk 1]
```diff
@@ -1,131 +1,88 @@
...
-	case aclGroup:
-		return fmt.Sprintf("group:%v:", a.getID())
-	case aclMask:
-		return "mask::"
-	case aclOthers:
-		return "other::"
-	}
-	return "?:"
-}
+	// Permissions.
```
Comment (MichaelEischer) [inline internal/dump/acl.go:10]: The trailing dot in the single-word comments looks a bit odd.

File: internal/dump/acl.go [hunk 2]
```diff
@@ -1,131 +1,88 @@
...
-	var finalacl string
-	for _, acl := range a.List {
-		finalacl += acl.String() + "\n"
+func aclPermText(p uint16) []byte {
+	s := []byte("---")
+	if p&aclPermRead != 0 {
+		s[0] = 'r'
+	}
+	if p&aclPermWrite != 0 {
+		s[0] = 'w'
```
Comment (MichaelEischer) [inline internal/dump/acl.go:82]: The offsets are wrong, for write it should be `1` and `2` for execute.

File: internal/dump/acl.go [hunk 3]
```diff
@@ -1,131 +1,88 @@
...
+	if len(acl)-4 < 0 || (len(acl)-4)%8 != 0 {
+		return "", errors.New("wrong length")
 	}
-	for binary.Read(nr, binary.LittleEndian, ae) == nil {
-		elem.aclSID = (aclSID(ae.Tag) << 32) | aclSID(ae.ID)
-		elem.Perm = ae.Perm
-		a.List = append(a.List, elem)
+	version := binary.LittleEndian.Uint32(acl)
+	if version != 2 {
+		return "", errors.New("wrong version")
```
Comment (MichaelEischer) [inline internal/dump/acl.go:39]: wouldn't `unsupported ACL version` be more accurate?

**Review by MichaelEischer — CHANGES_REQUESTED** [review]
The permission formatting seems to be wrong, other than that the PR is fine. I'd slightly prefer a certain separation in decoding and formatting the ACLs, but it's good enough for now.

**Comment by MichaelEischer** [comment]
@greatroar ping?

File: internal/dump/acl.go (hunk 1 above)
Comment (greatroar) [reply internal/dump/acl.go]: Fixed.

File: internal/dump/acl.go (hunk 2 above)
Comment (greatroar) [reply internal/dump/acl.go]: Fixed, tests amended (and tested again against an actual ACL).

File: internal/dump/acl.go (hunk 3 above)
Comment (greatroar) [reply internal/dump/acl.go]: Changed to "unsupported ACL format version".

**Review by MichaelEischer — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**